### PR TITLE
osd/scrub: only telling the scrubber of 'updates' events if these eve…

### PR DESCRIPTION
**Pull request by ronen-fr**
Only telling the scrubber of 'update' events if these events are awaited.

Only the Primary, and only when there is an active scrubbing going on and the
scrubber's state machine is in WaitUpdates states, should be notified of 'updates'.
The extra messages currently queued and processed are, apart from creating redundant
log lines and wasting CPU time, complicating the task of preventing a race in the handling
of stale scrubs.

Signed-off-by: Ronen Friedman <rfriedma@redhat.com>

### CODE REVIEW

**Comment by ronen-fr** [comment]
Passed minimal teuthology testing:
http://pulpito.front.sepia.ceph.com/?branch=wip-ronenf-last-update

File: src/osd/scrub_machine.cc [hunk 1]
```diff
@@ -60,6 +60,12 @@ bool ScrubMachine::is_reserving() const
   return state_cast<const ReservingReplicas*>();
 }
 
+bool ScrubMachine::is_accepting_updates() const
+{
+  // note that this specific test is Primary-specific
```
Comment (athanatos) [inline src/osd/scrub_machine.cc:65]: Replace comment with assert(is_primary());

**Comment by athanatos** [comment]
I don't really understand how this prevents a race.

File: src/osd/scrub_machine.cc (hunk 1 above)
Comment (ronen-fr) [reply src/osd/scrub_machine.cc:65]: Not in the interface currently available to the state machine.
I'd rather not add it just for this assert

**Comment by ronen-fr** [comment]
> I don't really understand how this prevents a race.

I am working on the PR that actually prevents the race (by introducing a flag that is On from the very first
queuing of the message that initiates the scrub until after the 'scrub finished' event is handled by the state machine).
But the original reason I created this (42714) PR, was because of a test failure where a stale 'update' message
received at the wrong time is what triggered the 2'nd scrub. Again - the real solution is in the next PR. But the
extra messages do complicate the understanding of the edge scenarios.

File: src/osd/scrub_machine.cc (hunk 1 above)
Comment (athanatos) [reply src/osd/scrub_machine.cc:65]: I think it's worth it.  Quite a few things in this portion of the code base rely on being queued only for replicas or primaries.  Here, you felt it was important enough to warrant a comment.  I think being able to tell dynamically if this constraint is violated is worth a bit of interface bloat.

**Review by athanatos — APPROVED** [review]
LGTM other than the assert(is_primary()) change.

File: src/osd/scrub_machine.cc (hunk 1 above)
Comment (ronen-fr) [reply src/osd/scrub_machine.cc:65]: Will fix.

**Comment by ronen-fr** [comment]
jenkins test make check

**Comment by ronen-fr** [comment]
jenkins test make check

**Comment by neha-ojha** [comment]
jenkins test make check

**Comment by ronen-fr** [comment]
fails "make check" in check-generated.sh, which does not seem to be related.

**Comment by ronen-fr** [comment]
jenkins retest this please

**Comment by tchaikov** [comment]
https://pulpito.ceph.com/kchai-2021-08-13_12:30:46-rados-wip-kefu-testing-2021-08-13-1658-distro-basic-smithi/

cephadm failures are tracked by https://tracker.ceph.com/issues/52279

**Pull request merged**